File: leetcode/c++/medium/_3152_special_array_2.cpp
// Leetcode Problem 3152

#include <gtest/gtest.h>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<bool> isArraySpecial(vector<int>& nums, vector<vector<int>>& queries) {
        unordered_map<int, int> positions;

        int section = 0;
        bool was_last_even = nums[0] % 2 == 0;

        positions.insert({0, 0});

        for (int i = 1; i < nums.size(); i++) {
            if ((was_last_even && nums[i] % 2 != 0) || (!was_last_even && nums[i] % 2 == 0)) {
                positions.insert({i, section});
            } else {
                section++;
                positions.insert({i, section});
            }
            was_last_even = nums[i] % 2 == 0;
        }

        vector<bool> solution;
        for (int i = 0; i < queries.size(); i++) {
            if (positions[queries[i][0]] != positions[queries[i][queries[i].size() - 1]]) {
                solution.push_back(false);
            } else {
                solution.push_back(true);
            }
        }

        return solution;
    }
};


class TestIsArraySpecial : public ::testing::Test {
protected:
    Solution solution;
};

TEST_F(TestIsArraySpecial, SimpleSingleQueryCase) {
    vector<int> nums = {3, 4, 1, 2, 6};
    vector<vector<int>> queries = {{0, 4}};
    vector<bool> expected = {false};

    EXPECT_EQ(solution.isArraySpecial(nums, queries), expected);
}

TEST_F(TestIsArraySpecial, SimpleMultipleQueriesCase) {
    vector<int> nums = {4, 3, 1, 6};
    vector<vector<int>> queries = {{0, 2}, {2, 3}};
    vector<bool> expected = {false, true};

    EXPECT_EQ(solution.isArraySpecial(nums, queries), expected);
}

File: leetcode/c#/medium/_3152_SpecialArray_2.cs
// Leetcode Problem 3152

using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace medium;

public class Solution_SpecialArray2 {
    public bool[] IsArraySpecial(int[] nums, int[][] queries) {
        Dictionary<int, int> positions = [];

        int section = 0;
        bool wasLasEven = nums[0] % 2 == 0;

        positions.Add(0, 0);

        for (int i = 1; i < nums.Length; i++) {
            if ((wasLasEven && nums[i] % 2 != 0) || (!wasLasEven && nums[i] % 2 == 0)) {
                positions.Add(i, section);
            } else {
                section++;
                positions.Add(i, section);
            }
            wasLasEven = nums[i] % 2 == 0;
        }

        bool[] solution = new bool[queries.Length];
        for (int i = 0; i < queries.Length; i++) {
            if (positions[queries[i][0]] != positions[queries[i][^1]]) {
                solution[i] = false;
            } else {
                solution[i] = true;
            }
        }

        return solution;
    }

    [TestClass]
    public class TestSpecialArray2 {
        private Solution_SpecialArray2 solution;

        [TestInitialize]
        public void Setup() {
            solution = new Solution_SpecialArray2();
        }

        [TestMethod]
        public void TestSpecialArray2_SimpleTestCase1() {
            int[] nums = [3, 4, 1, 2, 6];
            int[][] queries = [[0, 4]];
            bool[] expected = [false];

            CollectionAssert.AreEqual(expected, solution.IsArraySpecial(nums, queries));
        }

        [TestMethod]
        public void TestSpecialArray2_SimpleTestCase2() {
            int[] nums = [4, 3, 1, 6];
            int[][] queries = [[0, 2], [2, 3]];
            bool[] expected = [false, true];

            CollectionAssert.AreEqual(expected, solution.IsArraySpecial(nums, queries));
        }
    }
}